apache/superset
# [SQL Lab] Async query results serialization with MessagePack and PyArrow

### CATEGORY

Choose one


### SUMMARY

Async query performance in SQL Lab, particularly with large result sets, is fairly poor due to how data is serialized and stored. This PR introduces a `RESULTS_BACKEND_USE_MSGPACK` config option to use [PyArrow](https://arrow.apache.org/docs/python/) to serialize the pandas DataFrame directly, and [MessagePack](https://github.com/msgpack/msgpack-python) for serializing the results payload. Compared to the existing JSON serialization, Arrow and MessagePack provide improved performance and result in much smaller payloads sent to S3 or other cache backends.

### Benchmarks with 100K rows of `birth_names` examples data (multiple runs)

#### JSON

avg serialization: 573ms
avg deserialization: 200ms
total: 773ms
compressed payload size: 816718
peak memory usage: 281.1 MiB

#### Arrow/msgpack

avg serialization: 70ms
avg deserialization: 40ms
total: 110ms
compressed payload size: 452634
peak memory usage: 266.2 MiB

_Benchmarks were performed on a Macbook Pro 2.6 GHz i7, 32GB running macOS 10.14.5 and Python 3.6.8. [memory-profiler](https://pypi.org/project/memory-profiler/) was used for memory usage stats._

### TEST PLAN

Testing thus far has been limited to Postgres, mainly on Superset examples data. For full compatibility testing:
- Enable `RESULTS_BACKEND_USE_MSGPACK = True` in `superset_config.py`
- Run queries in SQL Lab with various DB backends containing multiple data types
- Ensure displayed results and CSV downloads contain correctly formatted data

### ADDITIONAL INFORMATION



### REVIEWERS
@mistercrunch
